### PR TITLE
adding support for right to left languages

### PR DESCRIPTION
**Pull request by pioneer2637**
Currently just supporting for Farsi language has been added. But  by updating the persian_lookup_table located in file lv_persian.h it can easily be ported for other right to left languages such as Arabic.
There is no attribute that should be set .the only thing needed is providing an appropriate font

### CODE REVIEW

**Comment by kisvegabor** [comment]
Hi,

Thank you for this great update!

Please add some formatting to your code:
1. Use 'lv_misc/lv_templ.c/h'
2. Add comments to the function as you can see in other places
3. For consistency use `lower_case_variable_names`
4. Read https://github.com/littlevgl/lvgl/wiki/Coding-Style-Guide

And please provide a short code example and example font to make the testing easier.

**Comment by amirgon** [comment]
What happens with strings that contain a mixture of RTL and LTR characters?  
This is usually handled by a [Bidirectional Algorithm](https://www.w3.org/International/articles/inline-bidi-markup/uba-basics), and Unicode has several [reference implementations](https://www.unicode.org/Public/PROGRAMS/BidiReferenceC/).

**Comment by pioneer2637** [comment]
> Hi,
> 
> Thank you for this great update!
> 
> Please add some formatting to your code:
> 
> 1. Use 'lv_misc/lv_templ.c/h'
> 2. Add comments to the function as you can see in other places
> 3. For consistency use `lower_case_variable_names`
> 4. Read https://github.com/littlevgl/lvgl/wiki/Coding-Style-Guide
> 
> And please provide a short code example and example font to make the testing easier.

Hi 
You're welcome

Sorry for not following the rules and coding styles
OK, I will modify them and commit again as soon as possible

**Comment by pioneer2637** [comment]
> What happens with strings that contain a mixture of RTL and LTR characters?
> This is usually handled by a [Bidirectional Algorithm](https://www.w3.org/International/articles/inline-bidi-markup/uba-basics), and Unicode has several [reference implementations](https://www.unicode.org/Public/PROGRAMS/BidiReferenceC/).

The library can handle it. For proper handling this situation at first I'v checked that does this part of String is RTL or LTR then according to that part of string the buffer would be copied in reversed order or without any changes

**Comment by pioneer2637** [comment]
This is an example to show how you can use persian and english at the same time by using the font provided in this library

void lv_tutorial_hello_world_farsi(void) {
	lv_style_t* style;
	lv_obj_t * label1 = lv_label_create(lv_scr_act(), NULL);
	style = lv_obj_get_style(label1);
	lv_font_add(&iran_sans_25_2, &iran_sans_25_1);
	lv_font_add(&iran_sans_25_3, &iran_sans_25_1);
	lv_font_add(&iran_sans_25_4, &iran_sans_25_1);
	style = lv_obj_get_style(label1);
	style->text.font = &iran_sans_25_1;
	lv_label_set_text(label1, "Hello World!\n سلام دنیا ");
	lv_obj_align(label1, NULL, LV_ALIGN_CENTER, 0, 0);
}

**Comment by amirgon** [comment]
@pioneer2637 Do you handle intermixed RTL and LTR on the same line? On your example these are on separate lines (with a newline separator). How do you handle neutral characters (without explicit LTR or RTL property) such as punctuations, exclamation mark etc.?

For example: 

    Hello 10 World! 20!

Should probably look like

Hello 10 World! 20!
    
Note that in this example the exclamation mark (`!`) is positioned differently depending whether it is on the end of a RTL sequence or LTR sequence.

**Comment by pioneer2637** [comment]
> @pioneer2637 Do you handle intermixed RTL and LTR on the same line? On your example these are on separate lines (with a newline separator). How do you handle neutral characters (without explicit LTR or RTL property) such as punctuations, exclamation mark etc.?
> 
> For example:
> 
> ```
> Hello 10 World! 20!
> ```
> Should probably look like
> 
> ![Hello 10 World! 20!]
> 
> Note that in this example the exclamation mark (`!`) is positioned differently depending whether it is on the end of a RTL sequence or LTR sequence.

I've just tested your example on my board and it was exactly as It was written in code.
For handling both RTL and LTR firstly I split the text buffer into chunks of smaller buffer and separate them as RTL or LTR then for RTL parts copy them in reversed order in new buffer but for LTR parts just copy them without reversing.
The point here is, this optimization works well for writing English words in middle of Farsi sentences not writing Farsi in middle of English words.

File: lv_misc/lv_persian.c [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+uint8_t* lv_get_reversed_buffer(uint8_t*buffer, uint16_t start, uint16_t end) {
+	struct text_block_s block;
+	struct text_block_s *current_block;
+	struct text_block_s *last_block;
+	bool contain_persian_letter = false;
+	if ((end - start) < 3) {
+		return buffer;
+	}
+	for (int i = start; i < end; i++) {
+		if ((*((uint16_t*) &buffer[i]) & 0xFC) == 0xD8) {
```
Comment (amirgon) [inline lv_misc/lv_persian.c:212]: How can this be generalized to other RTL languages? You only take into account Persian letters in order to set `contain_persian_letter`. How about a more general check that would set `contains_RTL_letter`?

File: lv_misc/lv_persian.c [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	uint8_t* reversed_buff = lv_mem_alloc(end - start + 2);
+	if(reversed_buff == NULL){
+		return buffer;
+	}
+	*((uint32_t*)&reversed_buff[end-start-2]) = NULL;
+	current_block = &block;
+	block.next = NULL;
+	block.start = start;
+	block.end = end;
+	for (int i = end - 1; i > start; i--) {
```
Comment (amirgon) [inline lv_misc/lv_persian.c:229]: Could you explain what you are doing? What is the purpose of these blocks you are allocating here and freeing at the end of the function? I suggest adding more documentation to the code.

File: lv_draw/lv_draw_label.c [hunk 3]
```diff
@@ -54,148 +52,214 @@ static uint8_t hex_char_to_num(char hex);
...
+			if ((const char*)reversed_buffer == txt) {
+				letter = lv_txt_encoded_next(txt, &i);
+				reversed_index = i;
+			} else {
+				letter = lv_txt_encoded_next((const char*) reversed_buffer,
+						&reversed_index);
+				current_letter = letter;
+				// update i
+				i = line_start + reversed_index;
+				if ((letter > 0x600) && (letter < 0x6ff)) {
```
Comment (amirgon) [inline lv_draw/lv_draw_label.c:161]: What is the range 0x600 - 0x6ff? Is that related to Persian letters encoding? Is there a way to generalize it to other RTL languages?

File: lv_draw/lv_draw_label.c [hunk 4]
```diff
@@ -54,148 +52,214 @@ static uint8_t hex_char_to_num(char hex);
...
+		uint32_t previous_letter;
+		uint32_t current_letter;
+		reversed_index = 0;
+		uint8_t noWidth = 0;
+		// end my code
+		while (i < line_end) {
+			// my code
+			//if it's necessary make the txt buffer reversed
+			if (!reversed_index) {
+				reversed_buffer = lv_get_reversed_buffer((uint8_t*) txt, i,
```
Comment (amirgon) [inline lv_draw/lv_draw_label.c:147]: `lv_get_reversed_buffer` will be called frequently, also when there is no Persian letter in the text. I think there should be a way to opt out and avoid this extra processing when developing an application exclusively for LTR languages. This can easily be done by a global `#define` that enables/disabled RTL processing.

**Review by amirgon — CHANGES_REQUESTED** [review]
I've added some comments related to this PR.  
I think the main issue here is that the code is specific to Persian script.  
I believe there should be a way to provide RTL support which is not language specific and can be extended to other languages such as Arabic, Hebrew etc.

File: lv_draw/lv_draw_label.c [hunk 5]
```diff
@@ -6,6 +6,7 @@
 /*********************
  *      INCLUDES
  *********************/
+#include <lv_misc/lv_persian.h>
```
[the remark this answers is not in the document]
Comment (pioneer2637) [reply lv_draw/lv_draw_label.c]: You're right.Know it just support Farsi because it is my native language and I'm totally familiar with it but based on what I've done here it can be easily adapted to other RTL languages. For supporting Arabic  language you just need to update the lookup table and provide appropriate codes for  the Arabic specific Letters. Those Codes can be extracted from  [here](https://utf8-chartable.de/unicode-utf8-table.pl) The UTF range from 0xFB00 to 0xFB4F are for Arabic and Farsi. For other RTL Languages you should find the right UTF range to convert those letters base on their position in a word.

File: lv_draw/lv_draw_label.c [hunk 6]
```diff
@@ -54,148 +52,214 @@ static uint8_t hex_char_to_num(char hex);
...
-        y_ofs = offset->y;
-        pos.y += y_ofs;
-    }
-
-    /*Real draw need a background color for higher bpp letter*/
+void lv_draw_label(const lv_area_t * coords, const lv_area_t * mask,
+		const lv_style_t * style, lv_opa_t opa_scale, const char * txt,
+		lv_txt_flag_t flag, lv_point_t * offset) {
+
+	// my code
```
[the remark this answers is not in the document]
Comment (pioneer2637) [reply lv_draw/lv_draw_label.c:62]: Sorry I forgot to clear them away. When I start writing codes I write those comments for myself to not confusing my codes with other part of codes.

File: lv_misc/lv_persian.c (hunk 1 above)
Comment (pioneer2637) [reply lv_misc/lv_persian.c:212]: If you look at this [link](https://utf8-chartable.de/unicode-utf8-table.pl) and choose the range between  U+0600...U+06FF you can see all Arabic and Farsi letter. When you are typing in Farsi the UTF8 code will be generated for you without considering different shape of letter for example when you type ج the letter code is 0xDA84 and this code is fix for all position of letters in a word. The rest is your job to determine which shape of letter should be placed here and this part could be extracted from range U+FB00 ... U+FB4F and that's how I create the lookup table.

File: lv_draw/lv_draw_label.c (hunk 3 above)
Comment (pioneer2637) [reply lv_draw/lv_draw_label.c:161]: This has been explained in previous comments

File: lv_draw/lv_draw_label.c (hunk 4 above)
Comment (pioneer2637) [reply lv_draw/lv_draw_label.c:147]: If I wanted to do this there must be an attribute that should be set for writing RTL and Right now I am working on a product that should support English and Farsi at the same time.
If you look at the function's body you would see this function only allocate buffer if it is necessary and if it is not necessary the pointer that was passed to  function will be returned without allocating new pointer. So I think this is not making considerable overhead for Processor

**Comment by kisvegabor** [comment]
Thank you for the improvements in style. I also pushed some minor fixes.

A few notes:
1. Please covert the Arabic font as built-in. It will make possible to enable/disable it in lv_conf.h
2. With `LV_TXT_RTL   0` no Latin letters are drawn

Unfortunately, I don't know Arabic languages to help the improvement in this regard, however, it's good to see that you are working on it! :+1:

**Comment by pioneer2637** [comment]
> Thank you for the improvements in style. I also pushed some minor fixes.
> 
> A few notes:
> 
> 1. Please covert the Arabic font as built-in. It will make possible to enable/disable it in lv_conf.h
> 2. With `LV_TXT_RTL   0` no Latin letters are drawn
> 
> Unfortunately, I don't know Arabic languages to help the improvement in this regard, however, it's good to see that you are working on it! 👍


You're welcome
I fixed the bug and about the Arabic language, right now I'm so busy but I'll do it as soon as possible

**Comment by embeddedt** [comment]
Would it be better to rename variables containing `persian` to contain `rtl`? The use of the name `persian` is misleading to me and makes it seem like it is not portable to other languages.

**Comment by kisvegabor** [comment]
@embeddedt I agree!

**Comment by pioneer2637** [comment]
> Would it be better to rename variables containing `persian` to contain `rtl`? The use of the name `persian` is misleading to me and makes it seem like it is not portable to other languages.

The reason that I've chosen persian instead of rtl is that I know Persian well and I know that this changes would work well with this language but I'm not sure about other RTL languages .If other people help me to complete this for other RTL Languages then I would change it as you suggested. I have provided the base methods and the procedure is the same for all other RTL languages.

**Comment by amirgon** [comment]
@embeddedt @kisvegabor 

> Would it be better to rename variables containing `persian` to contain `rtl`? The use of the name `persian` is misleading to me and makes it seem like it is not portable to other languages.

It's not only the string `persian` that is Persian specific. The code itself is not RTL generic, at least not the way it is written now. For example, there are hard-coded Persian-specific unicode ranges, not only in `lv_persian.c` but also in `lv_draw_label.c`

File: lv_draw/lv_draw_label.c (hunk 3 above)
Comment (amirgon) [reply lv_draw/lv_draw_label.c:161]: @pioneer2637,
My opinion is, that instead of specifying unicode ranges hard-coded in different places in your code, you could call a function (something like `bool lv_is_rtl_letter(uint32_t letter)`) and make the checks there.  
This would make porting to other languages easier. All we would have to do to support another language is to change `lv_is_rtl_letter` and add another range there.

Comment (kisvegabor) [reply lv_draw/lv_draw_label.c:161]: @amirgon It really sounds flexible. A default Persian `lv_is_rtl_letter` can be added to show an example.

**Comment by kisvegabor** [comment]
@pioneer2637 Have you closed the issue because you don't have time or don't want to improve this feature further?
